hummingbot/connector/exchange/kucoin/kucoin_api_order_book_data_source.py

```diff
@@ -145,74 +161,194 @@ async def ws_connect_data(self):
+        :param output: the output queue
+        """
+        all_symbols: List[str] = await self.get_trading_pairs()
+        all_symbols_set = set(all_symbols)
+
+        # removals
+        # remove any markets in current connections that are not present in the new master set
+        for task_index in self._tasks[stream_type]:
+            self._tasks[stream_type][task_index]["markets"] &= all_symbols_set
+
+        # additions
+        # from the new set of trading pairs, delete any items that are in the connections already
+        for task_index in self._tasks[stream_type]:
+            all_symbols_set -= self._tasks[stream_type][task_index]["markets"]
+
+        # now all_symbols_set contains just the additions, add each of those to the shortest connection list
+        for market in all_symbols_set:
+            smallest_index = 0
+            smallest_set_size = self.SYMBOLS_PER_CONNECTION + 1
+            for task_index in self._tasks[stream_type]:
+                if len(self._tasks[stream_type][task_index]["markets"]) < smallest_set_size:
+                    smallest_index = task_index
+                    smallest_set_size = len(self._tasks[stream_type][task_index]["markets"])
+            self._tasks[stream_type][smallest_index]["markets"].add(market)
+
+    def _stop_update_tasks(self, stream_type: StreamType):
+        if stream_type in self._tasks:
+            for task_index in self._tasks[stream_type]:
+                if not self._tasks[stream_type][task_index]["task"].done():
+                    self._tasks[stream_type][task_index]["task"].cancel()
+            del self._tasks[stream_type]
+
+    async def _outer_messages(self, stream_type: StreamType, task_index: int, output: asyncio.Queue):
         websocket_data: Dict[str, Any] = await self.ws_connect_data()
-        kucoin_ws_uri: str = websocket_data["data"]["instanceServers"][0]["endpoint"] + "?token=" + websocket_data["data"]["token"] + "&acceptUserMessage=true"
+        kucoin_ws_uri: str = websocket_data["data"]["instanceServers"][0]["endpoint"] + "?token=" + \
+            websocket_data["data"]["token"] + "&acceptUserMessage=true"
+        # connects and writes data to the output queue
         while True:
             try:
+                market_set: set = self._tasks[stream_type][task_index]["markets"]
                 async with websockets.connect(kucoin_ws_uri) as ws:
                     ws: websockets.WebSocketClientProtocol = ws
-                    for trading_pair in self._trading_pairs:
-                        subscribe_request: Dict[str, Any] = {
-                            "id": int(time.time()),
-                            "type": "subscribe",
-                            "topic": f"/market/level2:{trading_pair}",
-                            "response": True
-                        }
-                        await ws.send(json.dumps(subscribe_request))
+                    ping_task: asyncio.Task = \
```

Comment:
`ping_task` is used in `finally`, but its definition here depends on a successful `websockets.connect()`.

This is not good, and could lead to `_outer_messages()` exiting silently if `websockets.connect()` happens to fail.

What you should do instead, is to define `ping_task` as optional and initialize as None before the try statement. And in the finally block, check if `ping_task` is defined and cancel it there.